src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-unital, non-associative algebras
+
+This files defines non-unital, non-associative algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Main definitions
+
+  * `non_unital_non_assoc_algebra`
+  * `non_unital_non_assoc_algebra_hom`
+
+## Tags
+
+non-unital, non-associative, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra. -/
+@[protect_proj]
+class non_unital_non_assoc_algebra [non_unital_non_assoc_semiring A] extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.to_non_unital_non_assoc_semiring [semiring A] : non_unital_non_assoc_semiring A :=
+{ ..(infer_instance : semiring A) }
```

Comment:
This belongs in the other file.

Reply:
You can also write `‹semiring α›` rather than `(infer_instance : semiring A)` if you prefer.

Reply:
Thanks, done. Some resulting build failures which I will resolve shortly.